### DISCUSSION TITLE
KML Clamped to ground

### DISCUSSION THREAD

**Discussion by moisesrbd** [TerriaJS/terriajs#4902] [Q&A] [answered]
Hello friends
Is there any way to configure, in the init file catalog, that a KML item is drawn clamped to ground?
Cesium supports it ,but I don't know if Terria supports this type of configuration using, for example options = {} in the KML Item definition
Thanks.

**Comment by hardreddata** [comment]
You can `"clampToGround":true` in the catalog item. I am assuming this applies to the OGR converted layers. It is documented https://docs.terria.io/guide/connecting-to-data/catalog-type-details/geojson/ for geojson.

**Reply by moisesrbd** [reply]
HI, thanks for your answer.
I tested GeoJSON Item with clampToGround and works ok, but KML item still not working.
Im looking for something like this:
`        {
          "name": "Vic Police (KML)",
          "type": "kml",
          "url": "test/vic_police.kml",
          "clampToGround" : true
        }`

**Comment by hardreddata** [comment]
I actually struggle to understand this setting also.

I think if your data file includes an elevation component, switch to a 3D terrain map type, and if your data doesn't include a Z component, use the smooth globe.

Or maybe it is broken, or doesn't work for kml. Someone else will better be able to comment on this.

**Comment by tephenavies** [comment, marked as answer]
Hi @moisesrbd. Reading through KmlCatalogItem.js it looks like polygons should be clamped to the ground [here](https://github.com/TerriaJS/terriajs/blob/master/lib/Models/KmlCatalogItem.js#L256) whenever using 3D mode (terrain or smooth). There is no option like in GeoJsonCatalogItem to disable clamping to ground. What type of geometry do you have?

**Reply by tephenavies** [reply]
I've looked at the difference between KML and GeoJSON catalog items and I think KmlCatalogItem.js just needs to add `clampToGround` as a parameter (like [in GeoJsonCatalogItem ](https://github.com/TerriaJS/terriajs/blob/07a300e473475af188d81ea5656de6f7b73ce312/lib/Models/GeoJsonCatalogItem.js#L89) and [tracking updates to `clampToGround` ](https://github.com/TerriaJS/terriajs/blob/07a300e473475af188d81ea5656de6f7b73ce312/lib/Models/GeoJsonCatalogItem.js#L103)). Then change the signature of every call to `dataSource.load()` in KmlCatalogItem starting [here](https://github.com/TerriaJS/terriajs/blob/master/lib/Models/KmlCatalogItem.js#L160) to pass through an `options` object as the second parameter:

```js
var options = {
  sourceUri: proxyCatalogItemUrl(that, that.dataSourceUrl, "1d"),
  clampToGround: that.clampToGround
};
```

[Here are the docs for the `dataSource.load()` function in Cesium](https://cesium.com/docs/cesiumjs-ref-doc/KmlDataSource.html?classFilter=kmld#load). It looks like at the moment we aren't calling this with the right parameters, (possibly there was a breaking change to Cesium some time ago when clampToGround was added to Cesium).

Would you be interested in writing a PR for this change @moisesrbd?

**Reply by tephenavies** [reply]
And I think [`doneLoading`](https://github.com/TerriaJS/terriajs/blob/07a300e473475af188d81ea5656de6f7b73ce312/lib/Models/KmlCatalogItem.js#L252-L307) could be removed from KmlCatalogItem too since Cesium will take care of all clamping.

**Reply by moisesrbd** [reply]
Hi Steve.
Thanks for your good answer.
Indeed, I reviewed the implementation in Terriajs v8 (next branch) and made changes to make the KML clamped.
I did not want to publish it because I consider that it is a functionality with little use. Only that I needed it in special cases that we wanted to draw the KML on a 3DTile representing an open pit, without having to give the Z coordinates of each point of the polygons.
In the case of Terria next, I added the corresponding property to KmlCatalogItemTraits and Modified KmlCatalogItem to load in the same way as GeoJSONCatalogItem.

Thank you very much again.
Best 
Moises

**Reply by tephenavies** [reply]
We would welcome a PR to add `clampToGround` to KmlCatalogItem in `next`/v8. Actually I also just wrote [a ticket for improvements that we should make to KmlCatalogItem in `next`](https://github.com/TerriaJS/terriajs/issues/4958) after I wrote that answer up (thinking you were working off version 7). If you have some changes that could tick off one of those boxes that'd be great.
